crates/utils/re_video/src/stable_index_deque.rs

```diff
@@ -0,0 +1,322 @@
+use std::collections::VecDeque;
+
+/// A deque with an offset that allows indices into it to stay valid.
+///
+/// Does not expose the underlying [`VecDeque`] directly to guarantee that all operations are valid.
+/// This is useful to use stable indices for addressing in a growing/shrinking deque
+/// without having to resort to a more complex map datastructure.
+///
+/// To illustrate:
+/// ```
+/// // Indices for deque are unstable:
+/// # use std::collections::VecDeque;
+/// let mut v = (0..2).collect::<VecDeque<i32>>();
+/// assert_eq!(v.get(1), Some(&1));
+/// v.pop_front();
+/// assert_eq!(v.get(1), None);
+///
+/// // Indices for `StableIndexDeque` are stable:
+/// # use re_video::StableIndexDeque;
+/// let mut v = (0..2).collect::<StableIndexDeque<i32>>();
+/// assert_eq!(v.get(1), Some(&1));
+/// v.pop_front();
+/// assert_eq!(v.get(1), Some(&1));
+/// ```
+#[derive(Default, Clone, Debug)]
+pub struct StableIndexDeque<T> {
+    vec: VecDeque<T>,
+    index_offset: usize,
+}
+
+impl<T> StableIndexDeque<T> {
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            vec: VecDeque::new(),
+            index_offset: 0,
+        }
+    }
+
+    /// Creates a new deque from an iterator and an index offset.
+    ///
+    /// ```
+    /// # use re_video::StableIndexDeque;
+    /// let v = StableIndexDeque::from_iter_with_offset(0..2, 1);
+    /// assert_eq!(v.get(0), None);
+    /// assert_eq!(v.get(1), Some(&0));
+    /// assert_eq!(v.get(2), Some(&1));
+    /// ```
+    pub fn from_iter_with_offset(iter: impl IntoIterator<Item = T>, index_offset: usize) -> Self {
```

Comment:
This may be just me, but I feel the offset should come _first_, because it indicates the elements that are missing in the front:

`{2 ellided} + [a, b, c]`  ==  `[-, -, a, b, c]`

Reply:
yeah that makes sense! :)